src/bindgen/processor/block_overloads.cr

```diff
@@ -0,0 +1,71 @@
+module Bindgen
+  module Processor
+    # Generates overloads for methods that accept a single block argument.
+    # Those methods are ambiguous because Crystal does not permit overloading
+    # through different type restrictions on the block; this processor rectifies
+    # this by requiring each overload to also specify the types of the block
+    # parameters.  For example, the following code
+    #
+    # ```
+    # def f(& : Int32 ->) end
+    # def f(& : Bool ->) end
+    # def f(& : Int32, Bool ->) end
+    # ```
+    #
+    # is transformed into
+    #
+    # ```
+    # def f(_type1_ : Int32.class, & : Int32 ->) end
+    # def f(_type1_ : Bool.class, & : Bool ->) end
+    # def f(_type1_ : Int32.class, _type2_ : Bool.class, & : Int32, Bool ->) end
```

Comment:
Idea: The processor could detect a fourth case:
`def f(_type1_ : UInt64.class, _type2_ : Bool.class, & : UInt64, Bool ->)`.  For this case, the `_type2_` could default to `Bool.class`, as the ambiguity is already resolved by the first argument.

This could be done through a later PR however, as our use-case is fixed by this for now.

Reply:
First, the defaulting would be broken again by introducing another overload:

```crystal
def f(_type1_ : Int32.class, _type2_ : Bool.class, & : Int32, Bool ->) end
def f(_type1_ : UInt64.class, _type2_ : Bool.class, & : UInt64, Bool ->) end
def f(_type1_ : UInt64.class, _type2_ : Float64.class, & : UInt64, Float64 ->) end
```

Second, the defaulting wouldn't work either if there are other overloads taking the same type list without those omitted types:

```crystal
def f(_type1_ : Int32.class, & : Int32 ->)
  yield(1)
end
def f(_type1_ : UInt64.class, & : UInt64 ->)
  yield(2_u64)
end
def f(_type1_ : Int32.class, _type2_ : Bool.class = Bool, & : Int32, Bool ->)
  yield(3, false)
end
def f(_type1_ : UInt64.class, _type2_ : Bool.class = Bool, & : UInt64, Bool ->)
  yield(4_u64, false)
end

f(Int32, &Proc(Int32, Nil).new {|x| puts x}) # 3

f
# Error: wrong number of arguments for 'f' (given 0, expected 1..2)
#
# Overloads are:
#  - f(_type1_ : Int32.class, _type2_ : Bool.class = Bool, &)
#  - f(_type1_ : UInt64.class, _type2_ : Bool.class = Bool, &)
```

In fact Crystal _overwrites_ the first two overloads with the last two, and vice versa if the definition order is reversed. So I think having defaultable type arguments would make the processor overcomplicated with little applicability.